File: semana3.py
#=====================================================================
#Projeto semana 3
#Felipe Caruzzo Rosini e José Vinicio Fernandes da Costa
#=====================================================================
class Jogador:
    def __init__(self, nome="Sem nome", hp_maximo=100):
        self.__nome = nome
        self.__hp_maximo = hp_maximo
        self.__hp = hp_maximo
        self.__nivel = 1
        self.__xp = 0

    def __del__(self):
        print("Jogador", self.__nome, "foi destruido!")

    def get_nome(self):
        return self.__nome

    def get_nivel(self):
        return self.__nivel

    def set_nivel(self, nivel):
        if nivel >= 0:
            self.__nivel = nivel
        else:
            print("Nivel nao pode ser negativo")

    def get_hp_maximo(self):
        return self.__hp_maximo

    def get_hp(self):
        return self.__hp

    def get_xp(self):
        return self.__xp

    def set_hp(self, hp):
        if hp < 0:
            self.__hp = 0
        elif hp > self.__hp_maximo:
            self.__hp = self.__hp_maximo
        else:
            self.__hp = hp

    def exibir_status(self):
        print("Status do Jogador")
        print("Nome:", self.__nome)
        print("Nivel:", self.__nivel)
        print("Xp:", self.__xp, "/100")
        print("HP:", self.__hp, "/", self.__hp_maximo)
        if self.esta_vivo():
            print("Vivo: Sim")
        else:
            print("Vivo: Nao")

    def esta_vivo(self):
        return self.__hp > 0

    def receber_dano(self, dano):
        if dano > 0:
            self.set_hp(self.__hp - dano)
            print(self.__nome, "recebeu", dano, "de dano! HP atual:", self.__hp)
        else:
            print("Dano invalido")

    def curar(self, cura):
        if cura > 0:
            if self.esta_vivo():
                self.set_hp(self.__hp + cura)
                print(self.__nome, "foi curado em", cura, "HP atual:", self.__hp)
            else:
                print("Jogador morto, nao pode curar")
        else:
            print("Cura invalida")

    def subir_nivel(self):
        self.__nivel += 1
        self.__xp = 0

    def experiencia(self, quantidade):
        self.__xp += quantidade
        while self.__xp >= 100:
            self.__nivel += 1
            self.__xp -= 100

def main():
    heroi = Jogador("Henry", 100)

    heroi.exibir_status()
    heroi.receber_dano(30)
    heroi.curar(20)
    heroi.receber_dano(10)
    heroi.experiencia(110)
    heroi.exibir_status()
    heroi.curar(50)

if __name__ == "__main__":
    main()

File: semana3.cpp
//=====================================================================    
//Projeto semana 3
//Felipe Caruzzo Rosini e José Vinicio Fernandes da Costa
//=====================================================================    

#include <iostream>
#include <string>

using namespace std;

class Jogador {
private:
    string nome;
    int nivel;
    int hp;
    int hpMaximo;
    int xp;
    
public:
    Jogador(string n, int hpm) {
        nome = n;
        nivel = 1;
        hpMaximo = hpm;
        hp = hpm;
        xp = 0;
    }

    Jogador(string n) {
        nome = n;
        nivel = 1;
        hpMaximo = 100;
        hp = 100;
        xp = 0;
    }

    Jogador() {
        nome = "Sem nome";
        nivel = 1;
        hpMaximo = 100;
        hp = 100;
        xp = 0;
    }

    ~Jogador() {
        cout << "Jogador " << nome << " foi destruido!" << endl;
    }

    //========GETERS========
    string getNome() { 
        return nome; 
    }
    
    int getHpMaximo() { 
        return hpMaximo; 
    }
    
    int getNivel() { 
        return nivel; 
    }
    
    int getHp() { 
        return hp; 
    }

    int getXp(){
        return xp;
    }


    //========SETERS========
    void setNivel(int n) {
        if (n >= 0) {
            nivel = n;
        } else {
            cout << "Erro: Nivel nao pode ser negativo" << endl;
        }
    }

    void setHp(int h) {
        if (h < 0) {
            hp = 0;
        } else if (h > hpMaximo) {
            hp = hpMaximo;
        } else {
            hp = h;
        }
    }

    void setXp(int n){
        this->xp = n;
    }


    //=======MÉTODOS=========
    void exibirStatus() {
        cout << "Status do Jogador" << endl;
        cout << "Nome: " << nome << endl;
        cout << "Nivel: " << nivel << endl;
        cout << "Xp: " << xp << "/100" << endl;
        cout << "HP: " << hp << "/" << hpMaximo << endl;
        if (estaVivo()) {
            cout << "Vivo: Sim" << endl;
        } else {
            cout << "Vivo: Nao" << endl;
        }
    }

    bool estaVivo() {
        return hp > 0;
    }

    void receber_dano(int quantidade) {
        if (quantidade > 0) {
            setHp(hp - quantidade);
            cout << nome << " recebeu " << quantidade << " de dano! HP atual: " << hp << endl;
        } else {
            cout << "Dano invalido" << endl;
        }
    }

    void curar(int quantidade) {
        if (quantidade > 0) {
            if (estaVivo()) {
               setHp(hp + quantidade);
               cout << nome << " curou " << quantidade << " de HP atual: " << hp << endl;
            } else {
                cout << "Jogador morto, nao pode ser curado" << endl;
            }
        } else {
            cout << "Cura invalida" << endl;
        }
    }

    void subirNivel(){
        nivel = nivel + 1;
        xp = 0;
    }

    void experiencia (int quantidade) {
        xp = xp + quantidade;
        while (xp >= 100){
            nivel = nivel + 1;
            xp = xp - 100;
        }
    }
    //=========FIM MÉTODOS=========
};

class armas{
private:
    string nome;
    float desgate;

public:
    armas(string nome){
        this->nome = nome;
        this->desgate = 100;
    }

    void setNome(string Nome){
        this->nome = Nome;
    }
    
    void setDesgaste(float Desgaste){
        this->desgate = Desgaste;
    }

    string getNome(){
        return this->nome;
    }

    float getDesgaste(){
        return this->desgate;
    }

    ~armas(){
        cout << this->getNome() << " foi destruido!" << endl;
    }
};

/*
class armasBrancas : public armas{
private:

public:

};

class armasFogo : public armas{
private:

public:

};

/*
class Policial : public Jogador{
private:
    
public:

};
*/
int main() {
    //Policial heroi("Henry", 100);
    armas a("Pistola");
    cout << a.getNome() << "|" << a.getDesgaste() << endl;

    return 0;
}
